bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java

```diff
@@ -0,0 +1,520 @@
+import org.eclipse.jdt.core.dom.QualifiedName;
+import org.eclipse.jdt.core.dom.SimpleType;
+import org.eclipse.jdt.internal.corext.dom.ASTNodes;
+import org.eclipse.jdt.ui.text.java.IInvocationContext;
+import org.eclipse.jdt.ui.text.java.IJavaCompletionProposal;
+import org.eclipse.jdt.ui.text.java.IProblemLocation;
+import org.eclipse.jdt.ui.text.java.IQuickFixProcessor;
+import org.eclipse.jface.text.contentassist.IContextInformation;
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.swt.graphics.Point;
+import org.osgi.framework.namespace.BundleNamespace;
+import org.osgi.framework.namespace.PackageNamespace;
+import org.osgi.resource.Capability;
+import org.osgi.resource.Requirement;
+import org.osgi.service.repository.Repository;
+
+import aQute.bnd.build.Project;
+import aQute.bnd.build.Workspace;
+import aQute.bnd.build.WorkspaceRepository;
+import aQute.bnd.build.model.BndEditModel;
+import aQute.bnd.build.model.clauses.VersionedClause;
+import aQute.bnd.header.Attrs;
+import aQute.bnd.osgi.resource.CapReqBuilder;
+import aQute.bnd.properties.Document;
+import aQute.bnd.properties.IDocument;
+import aQute.bnd.service.RepositoryPlugin;
+import aQute.lib.collections.MultiMap;
+import aQute.lib.io.ByteBufferInputStream;
+import aQute.lib.io.IO;
+import aQute.lib.strings.Strings;
+import bndtools.central.Central;
+import bndtools.central.RepositoryUtils;
+
+public class ImportPackageQuickFixProcessor implements IQuickFixProcessor {
+
+    private static final ILogger logger = Logger.getLogger(ImportPackageQuickFixProcessor.class);
+
+    // Relevance constants
+    final public static int ADD_BUNDLE = 15;
+    final public static int ADD_BUNDLE_WORKSPACE = ADD_BUNDLE + 1;
+
+    @Override
+    public boolean hasCorrections(ICompilationUnit unit, int problemId) {
+
+        switch (problemId) {
+            case IProblem.ImportNotFound :
+            case IProblem.UndefinedType :
+                return true;
+            default :
+                return false;
+        }
+    }
+
+    static private Requirement buildRequirement(String importName) {
+        String filter = String.format("(%s=%s)", PackageNamespace.PACKAGE_NAMESPACE, importName.trim());
+
+        return new CapReqBuilder(PackageNamespace.PACKAGE_NAMESPACE).addDirective(PackageNamespace.REQUIREMENT_FILTER_DIRECTIVE, filter)
+            .buildSyntheticRequirement();
+    }
+
+    private Set<Capability> searchRepository(Repository osgiRepo, String pkg) {
+
+        Requirement req = buildRequirement(pkg);
+        Set<Capability> resultSet = new LinkedHashSet<>();
+        Map<Requirement, Collection<Capability>> providers = osgiRepo.findProviders(Collections.singleton(req));
+
+        for (Collection<Capability> values : providers.values()) {
+            if (values == null) {
+                continue;
+            }
+
+            for (Capability providerCap : values) {
+                resultSet.add(providerCap);
```

Comment:
Thanks for highlighting this. I actually managed to simplify it even further by changing `searchRepositories()` to return a `Collection` rather than a `Set`, so that I can return the search results directly. This eliminates the copy completely and is this more efficient.